include/solvers/PreconditionCG.hpp

```diff
@@ -0,0 +1,256 @@
+// Copyright (c)   2017 Tyler Olsen
+//                 2018 Patrick Diehl
+//                 2019 Nanmiao Wu
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BLAZE_ITERATIVE_PRECONDITIONCG_HPP
+#define BLAZE_ITERATIVE_PRECONDITIONCG_HPP
+
+#include "PreconditionCGTag.hpp"
+
+BLAZE_NAMESPACE_OPEN
+ITERATIVE_NAMESPACE_OPEN
+
+    namespace detail {
+        // Decompose A as A = L + D + U, where L is strictly lower triangular matrix, D is diagonal matrix, and U is
+        // strictly upper triangular matrix.
+        template<typename MatrixType, typename T>
+        void preconditioner_matrix(std::string type, const MatrixType &A, MatrixType &M){
+
+            if (type.compare("Jacobi preconditioning")==0){
+                // M = D
+                std::size_t n = A.columns();
+                MatrixType D(n,n);
+                for(std::size_t i = 0; i < n; ++i){
+                    for(std::size_t j = 0; j < n; ++j){
+                        if(j == i){
+                            D(i,j) = A(i,j);
+                        }
+                        else {
+                            D(i,j) = 0;
+                        }
+                    }
+                }
+                M = D;
+            }
+
+            if (type.compare("Gauss Seidel preconditioning")==0){
+                std::size_t n = A.columns();
+                MatrixType L(n,n), D(n,n);
+                for(std::size_t i = 0; i < n; ++i){
+                    for(std::size_t j = 0; j < n; ++j){
+                        if(j < i){
+                            L(i,j) = A(i,j);
+                        }
+                        else{
+                            L(i,j) = 0;
+                        }
+                    }
+                }
```

Comment:
Couldn't the same be achieved using `blaze::LowerMatrix` (https://bitbucket.org/blaze-lib/blaze/wiki/Triangular%20Matrices)?

Reply:
Since A is not a StrictlyLowerMatrix, we cannot do it. A solution is to use blaze::band and a for loop.